File: juros.js
// Cálculo do juros, sendo que precisa de arrays pra isso
// Versão 0.1: 26/01/2025: versão de testes a partir da de JavaScript

// classe com os valores globais para simplificar as chamadas
class Juros{
	constructor(quantidade=0, composto=false, periodo=30){
		this.Quantidade = quantidade
		this.Composto = composto
		this.Periodo = periodo
		this.Pagamentos = new Array(0)
		this.Pesos = new Array(0)
	}

	// "Define as datas de pagamento a partir de uma string separada pelo delimitador
	setPagamentos(delimitador=",", pagamentos=""){
		if( pagamentos == ""){
			for(let c = 0; c < this.Quantidade; c++){
				this.Pagamentos[c] = (1 + c) * this.Periodo
			}
		} else {
			let temporaria = pagamentos.split(delimitador)
			for(let c = 0; c < this.Quantidade; c++){
				if(isNaN(Number(temporaria[c]))) return false
				this.Pagamentos[c] = Number(temporaria[c])
			}
		}
		return true
	}

	// "Define as datas de pagamento a partir de uma string separada pelo delimitador
	setPesos(delimitador=",", pesos=""){
		if( pesos == ""){
			for(let c = 0; c < this.Quantidade; c++){
				this.Pesos[c] = 1
			}
		} else {
			let temporaria = pesos.split(delimitador)
			for(let c = 0; c < this.Quantidade; c++){
				if(isNaN(Number(temporaria[c]))) return false
				this.Pesos[c] = Number(temporaria[c])
			}
		}
		return true
	}

	// Retorna a soma total de todos os pesos
	getPesoTotal(){
		let acumulador = 0
		for(let c = 0; c < this.Quantidade; c++){
			acumulador += this.Pesos[c]
		}
		return acumulador
	}

	// Calcula o acréscimo a partir dos juros
	jurosParaAcrescimo(juros=0){
		if(juros == 0 || this.Quantidade == 0 || this.Periodo == 0) return 0

		let total = this.getPesoTotal()
		let acumulador = 0

		for(let i = 0; i < this.Quantidade; i++){
			if(this.Composto){
				acumulador += this.Pesos[i] / ((1 + juros / 100) ** (this.Pagamentos[i] / this.Periodo))
			} else{
				acumulador += this.Pesos[i] / (1 + juros / 100 * this.Pagamentos[i] / this.Periodo)
			}
		}

		if (acumulador <= 0) return 0
		return (total / acumulador - 1) * 100
	}

	// Calcula os juros a partir do acréscimo
	acrescimoParaJuros(acrescimo=0, precisao=15, maximoInteracoes=100, maximoJuros=50, acrescimoComoValorOriginal=false){
		if(maximoInteracoes < 1 || this.Quantidade < 1 || precisao < 1 || this.Periodo < 1 || acrescimo <= 0 ) return 0

		let minimoJuros = 0
		let medioJuros = 0
		let total = this.getPesoTotal()

		if(total == 0) return 0

		if(acrescimoComoValorOriginal){
			acrescimo = 100 * (total / acrescimo - 1)
			if(acrescimo <= 0) return 0
		}

		var minimaDiferenca = 0.1 ** precisao

		for(let i = 0; i < maximoInteracoes; i++){
			medioJuros = (minimoJuros + maximoJuros) / 2
			if((maximoJuros - minimoJuros) < minimaDiferenca) return medioJuros
			if(this.jurosParaAcrescimo(medioJuros) <= acrescimo){
				minimoJuros = medioJuros
			} else{
				maximoJuros = medioJuros
			}
		}

		return medioJuros
	}
}

// cria o objeto juros da classe Juros e define os valores
let juros = new Juros(3, true, 30.0)
juros.setPagamentos(",", "30,60,90")
juros.setPesos(",", "1,1,1")

// calcula e guarda os resultados das funções
let pesoTotal = juros.getPesoTotal()
let acrescimoCalculado = juros.jurosParaAcrescimo(3.0)
let jurosCalculado = juros.acrescimoParaJuros(acrescimoCalculado, 18, 100, 50.0, false)

// imprime os resultados
console.log( "Peso total = " + pesoTotal)
console.log( "Acréscimo = " + acrescimoCalculado)
console.log( "Juros = " + jurosCalculado)

File: juros.cc
// Cálculo dos juros, sendo que precisa de parcelas pra isso
// Versão 0.1: 13/04/2026: feita sem muito conhecicmento de JADE

#include <jade.hpp>
#include <math.h>
#include <iomanip>
#include <sstream>

// algumas constantes para deixar o estilo um pouco mais coerente
#define POW pow
#define DOUBLE double
#define BOOL bool

// essa função é extremamente C++, mas é porque PRINT só imprime 5 casas, e não encontrei como resolver isso
SUB IMPRIME(STRING legenda, DOUBLE valor) DO
    std::ostringstream temp;
    temp << std::fixed << std::setprecision(15) << valor;
    PRINT(legenda, temp.str());
END

// variáveis globais, para simplificar as chamadas às funções e incializa escalares
INTEGER Quantidade = 3;
BOOL Composto = TRUE;
DOUBLE Periodo = 30.0;
ARRAY <DOUBLE> Pagamentos(Quantidade);
ARRAY <DOUBLE> Pesos(Quantidade);

// calcula a somatória dos elementos do array Pèsos[]
FUNCTION DOUBLE getPesoTotal() DO
    VAR acumulador = 0.0;

    FOR(VAR indice IN RANGE(0, Quantidade)) DO
        acumulador += Pesos[indice];
    NEXT

    RETURN acumulador;
END

// calcula o acréscimo a partir dos juros e parcelas
FUNCTION DOUBLE jurosParaAcrescimo(DOUBLE juros) DO
    VAR pesoTotal = getPesoTotal();
    IF (Quantidade < 1 OR Periodo <= 0.0 OR pesoTotal <= 0.0 OR juros <= 0.0) THEN RETURN 0.0; END
    VAR acumulador = 0.0;

    FOR(VAR indice IN RANGE(0, Quantidade)) DO
        IF (Composto) THEN
            acumulador += Pesos[indice] / POW(1.0 + juros / 100.0, Pagamentos[indice] / Periodo);
        ELSE
            acumulador += Pesos[indice] / (1.0 + juros / 100.0 * Pagamentos[indice] / Periodo);
        END
    NEXT

    IF (acumulador <= 0.0) THEN RETURN 0.0; END
    RETURN (pesoTotal / acumulador - 1.0) * 100.0;
END

// calcula os juros a partir do acréscimo e parcelas
FUNCTION DOUBLE acrescimoParaJuros(DOUBLE acrescimo, INTEGER precisao, INTEGER maxIteracoes, DOUBLE maxJuros) DO
    VAR pesoTotal = getPesoTotal();
    IF (Quantidade < 1 OR Periodo <= 0.0 OR pesoTotal <= 0.0 OR acrescimo <= 0.0 OR precisao < 1 OR maxIteracoes < 1 OR maxJuros <= 0.0) THEN RETURN 0.0; END
    VAR minJuros = 0.0, medJuros = maxJuros / 2.0, minDiferenca = POW(0.1, precisao);

    FOR(VAR iteracao IN RANGE(0, maxIteracoes)) DO
        IF (maxJuros - minJuros < minDiferenca) THEN RETURN medJuros; END
        IF (jurosParaAcrescimo(medJuros) < acrescimo) THEN minJuros = medJuros; ELSE maxJuros = medJuros; END
        medJuros = (minJuros + maxJuros) / 2.0;
    NEXT

    RETURN medJuros;
END

// função principal
MAIN
    // inicializa os elementos dos arrays Pagamentos[] e Pesos[]
    FOR(VAR indice IN RANGE(0, Quantidade)) DO
        Pagamentos[indice] = (indice + 1) * Periodo;
        Pesos[indice] = 1.0;
    NEXT

    // calcula e guarda os resultados das funções
    VAR pesoTotal = getPesoTotal();
    VAR acrescimoCalculado = jurosParaAcrescimo(3.0);
    VAR jurosCalculado = acrescimoParaJuros(acrescimoCalculado, 15, 65, 50.0);

    // imprime os resultados
    IMPRIME("Peso total = ", pesoTotal);
    IMPRIME("Acréscimo = ", acrescimoCalculado);
    IMPRIME("Juros = ", jurosCalculado);
END
